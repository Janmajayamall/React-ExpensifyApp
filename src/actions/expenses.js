import uuid from 'uuid';
import database from './../firebase/firebase.js';
import { get } from 'http';

//Actions for Expenses 
export const addExpense = (expense) => ({
    type:'ADD_EXPENSE',
    expense: expense
});

export const startAddExpense = (expenseData = {}) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
            } = expenseData;
   
   const expense = {description, note, amount, createdAt};
          
   return database.ref(`users/${uid}/expenses`).push(expense).then((ref)=>{
     dispatch(addExpense({
         id: ref.key,
         ...expense
     }));
   });
    };
};

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id: id
});

export const startRemoveExpense = (id) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense(id));
        });
    }
}

export const editExpense = (id, update)=>({
    type: 'EDIT_EXPENSE',
    id:id,
    update: update
});

export const startEditExpense = (id, update) =>{
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(update).then(()=>{
            dispatch(editExpense(id, update));
        });
    }
};



export const setExpenses = (expenses) =>({
    type: 'SET_EXPENSES',
    expenses: expenses
})

export const startSetExpenses = () => {
    return (dispatch, getState) =>{
        const uid = getState().auth.uid;
       return database.ref(`users/${uid}/expenses`).once("value").then((snapshot)=>{
            const expenses = [];

            snapshot.forEach((childSnapshot)=>{
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setExpenses(expenses));
        });
    };
};