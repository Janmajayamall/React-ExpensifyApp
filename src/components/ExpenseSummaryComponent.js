import React from 'react';
import getTotalExpenses from './../selectors/expensesTotal.js';
import numeral from 'numeral'


export default class ExpenseSummaryComponent extends React.Component{
    render(){
        const expenseSING = this.props.expenses.length === 1 ? 'expense' : 'expenses';
        return(
            <div>
                <p>Viewing {this.props.expenses.length} {expenseSING} totalling {numeral(getTotalExpenses(this.props.expenses)/100).format('$0,0.00')}</p> 
            </div>
        );
    }
}

