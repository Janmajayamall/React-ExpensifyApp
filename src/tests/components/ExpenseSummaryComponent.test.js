import React from 'react';
import {shallow} from 'enzyme';
import ExpenseSummaryComponent from './../../components/ExpenseSummaryComponent';
import expenses, {expenseSingle} from './../fixtures/expenses.js';

test('should render expense with 1 expense properly', ()=>{
    const wrapper = shallow(<ExpenseSummaryComponent expenses={expenseSingle}/>);
    expect(wrapper).toMatchSnapshot();

});

test('should render mutiple expenses properly', ()=>{
    const wrapper = shallow(<ExpenseSummaryComponent expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});