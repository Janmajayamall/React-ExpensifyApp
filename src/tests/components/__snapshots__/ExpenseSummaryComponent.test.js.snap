// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense with 1 expense properly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $0.02
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render mutiple expenses properly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totalling 
      <span>
        $0.03
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
